Nginx: Compile dynamic modules

TangRufus opened the pull request:
Note:
- it doesn't compile or install Nginx
- it assumes dynamic module sources are git repos
- it recompiles if Nginx version or dynamic module repos changed
- it skips when `nginx_dynamic_modules` is empty

Usage (see #867): 
```yaml
# group_vars/all/main.yml
nginx_dynamic_modules:
  ngx_brotli:
    repo: https://github.com/google/ngx_brotli.git
    objs:
      - ngx_http_brotli_filter_module.so
      - ngx_http_brotli_static_module.so
```

Related: #863

partounian commented on roles/nginx/defaults/main.yml:
```diff
@@ -15,3 +15,8 @@ nginx_cache_path: /var/cache/nginx
 nginx_cache_inactive: 1h
+
+# Dynamic modules
+nginx_deb_src: 'deb-src http://ppa.launchpad.net/nginx/development/ubuntu xenial main'
```
PPA is already set to this at the top of this file.

TangRufus replied:
How come you don't get this error?
```bash
$ sudo apt-get source nginx
Reading package lists... Done
E: You must put some 'source' URIs in your sources.list
```

partounian replied:
Sorry I didn't actually test these.

TangRufus replied:
Seems `$ add-apt-repository --enable-source {{ nginx_ppa }}` is a better choice

TangRufus commented:
Updated:
- `.so` files should have `644` file permission

TangRufus commented:
Will this be consider to be merged, or should I extract it to be a galaxy role?

Thanks!

swalkinshaw commented:
Still undecided on this one. But I'm on vacation so not really thinking about this 😉

TangRufus commented:
Closing this and continue on #867 because this seems won't be merged without `ngx_brotli`

swalkinshaw commented:
I was actually thinking more about this last night and I'm probably fine with it :)

I'll re-open and review it soon

TangRufus commented:
More changes to lure you into it:
 - rebased with 5e000e8
 - make use of #859 
 - don't copy `.so` files
 - removed `tags: nginx-dynamic-modules`
 - allow specific git version

TangRufus commented:
These commands works on my machine, not sure why it fails on travis:

```bash
$ ansible-playbook --syntax-check -e env=development deploy.yml
$ ansible-playbook --syntax-check -e env=development dev.yml
$ ansible-playbook --syntax-check -e env=development server.yml
```

strarsis commented:
@TangRufus: Can I use this for building the nginx pagespeed extension, too?

TangRufus commented:
> If you would like to build ngx_pagespeed as a dynamic module instead, use --add-dynamic-module= instead of --add-module=........
> [Pagespeed Doc](https://www.modpagespeed.com/doc/build_ngx_pagespeed_from_source)

I had trouble setting ngx_pagespeed as a dynamic module last August. It was a ngx_pagespeed  issue. But from the latest doc, it seems working now. You can give it a try.

---

This is untested and don't expect it will work without debugging:

```
# group_vars/all/main.yml
nginx_dynamic_modules:
  ngx_pagespeed:
    repo: https://github.com/apache/incubator-pagespeed-ngx.git
    version: some_comment_hash
    objs:
      - ngx_pagespeed.so
```

runofthemill commented:
this would be great to get merged - would allow Trellis to use https://github.com/masonicboom/ipscrub to reduce the surface area for GPDR issues

strarsis commented:
Can https://github.com/ANXS/nginx be somehow re-used for this?

runofthemill commented:
@strarsis that looks like it'd require a major refactor, since currently nginx is built & configured without an `ansible-galaxy` role, so highly unlikely, imo. did you try this PR with TangRufus' snippet to see if it works?
